### PR TITLE
93 pyqode alt

### PR DESCRIPTION
**Pull request by grace-harper-ibm**
### What are the issues this pull addresses (issue numbers / links)?
#93 

### Did you add tests to cover your changes (yes/no)?
No 

### Did you update the documentation accordingly (yes/no)?
Yes

### Did you read the CONTRIBUTING document (yes/no)?
Yes

### Summary
Removed Metal IDE functionality (PyQode) 
Adds "Developer Mode" which means a  filewatcher will alert the user if any of the QComponent source files are edited while the GUI is running 

### Details and comments
Anyone using an environment created using PyQode should have no trouble using Jupyter Notebook or even running tests in the terminal.
However, even though the code produces correct output, being run in the terminal with a pyqode-installed environment can cause `"Error processing line 1 of ~/opt/anaconda3/envs/qf392/lib/python3.7/site-packages/pyqode.qt-nspkg.pth:"` to be continuously thrown meaning perhaps this PR should be saved for a next release. We could consider adding a `develop` branch alongside main.

### CODE REVIEW

**Review by JeremyDrysdale-ibm — CHANGES_REQUESTED** [review]
The `Deep Dive` tutorial is failing on `gui.rebuild()`
image

File: qiskit_metal/_gui/widgets/qlibrary_display/delegate_qlibrary.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    """
+
+    def __init__(self, parent=None):
+        """
+        source_model_type - The Delegate may belong to a view using a ProxyModel.
+        However, the source model for that Proxy Model(s) should be a QFileSystemLibraryModel
+        is_dev_mode - Whether the MetalGUI is in Developer Mode or not
+
+        Args:
+            parent: parent
```
Comment (JeremyDrysdale-ibm) [inline qiskit_metal/_gui/widgets/qlibrary_display/delegate_qlibrary.py:40]: Throughout this file"
- Update docstring to include the return type 
- Update args to include argument types
- Update docstring to include exceptions

File: qiskit_metal/_gui/widgets/qlibrary_display/proxy_model_qlibrary.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        # (QComponent files) OR (Directories)
+        self.accepted_files__regex = r"(^((?!\.))(?!__init__)(?!_template)(?!__pycache__).*\.py)|(?!__pycache__)(?!base)(^([^.]+)$)"  # pylint: disable=line-too-long
+        self.setFilterRegExp(self.accepted_files__regex)
+        self.is_dev_mode = False
+
+    def set_dev_mode(self, ison: bool):
+        """
+        Set dev mode
+        Args:
+            ison: Whether dev mode is on
```
Comment (JeremyDrysdale-ibm) [inline qiskit_metal/_gui/widgets/qlibrary_display/proxy_model_qlibrary.py:44]: Throughout this file"
- Update docstring to include the return type 
- Update args to include argument types

File: qiskit_metal/_gui/widgets/qlibrary_display/tree_view_qlibrary.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        """
+        Args:
+            parent (QtWidgets.QWidget): parent widget
+        """
+        QTreeView.__init__(self, parent)
+        self.is_dev_mode = False  # Whether MetalGUI is in developer mode
+
+    def set_dev_mode(self, ison: bool):
+        """ Sets dev mode for self, model, model's source model, and delegate """
+        self.is_dev_mode = ison
```
Comment (JeremyDrysdale-ibm) [inline qiskit_metal/_gui/widgets/qlibrary_display/tree_view_qlibrary.py:43]: Throughout this file:

- Update docstring to include the return type
- Update args to include argument types
- Update docstring to include exceptions

File: qiskit_metal/designs/design_base.py [hunk 4]
```diff
@@ -447,37 +417,56 @@ def rebuild(self):  # remake_all_components
...
+        Reload the module and class of a given component and updates
+        all class instances. Then rebuilds all QComponents of that class
 
         Arguments:
-            component_module_name (str): String name of the module name, such as
-                `qiskit_metal.qlibrary.qubits.transmon_pocket`
-            component_class_name (str): String name of the class name inside thst module,
-                such  as `TransmonPocket`
+            qis_abs_path: Absolute to the QComponent source file to be reloaded
+
```
Comment (JeremyDrysdale-ibm) [inline qiskit_metal/designs/design_base.py:427]: Update docstring to include exceptions

**Review by JeremyDrysdale-ibm — CHANGES_REQUESTED** [review]

**Comment by priti-ashvin-shah-ibm** [comment]
> The `Deep Dive` tutorial is failing on `gui.rebuild()`
> image


On Crum, I am getting the same error with other notebook, tutorials/8 Full Design Flow Examples/Full_Chip_Design_01.ipynb

**Comment by JeremyDrysdale-ibm** [comment]
> > The `Deep Dive` tutorial is failing on `gui.rebuild()`
> > image
> 
> On Crum, I am getting the same error with other notebook, tutorials/8 Full Design Flow Examples/Full_Chip_Design_01.ipynb

That error will occurs anywhere `rebuild` is called.

File: qiskit_metal/_gui/main_window.py [hunk 5]
```diff
@@ -251,8 +231,8 @@ def __init__(self, design: QDesign = None):
         from .utility._handle_qt_messages import QtCore, _qt_message_handler
         QtCore.qInstallMessageHandler(_qt_message_handler)
 
-        qApp = kick_start_qApp()
-        if not qApp:
+        self.qApp = kick_start_qApp()
```
Comment (priti-ashvin-shah-ibm) [inline qiskit_metal/_gui/main_window.py:234]: Nice

File: qiskit_metal/designs/design_base.py [hunk 6]
```diff
@@ -447,37 +417,56 @@ def rebuild(self):  # remake_all_components
...
+        try:
+            # split on os.sep and / because PySide appears to sometimes use / on certain Windows
+            # Windows users' qis_abs_path may use os.sep or '/' due to PySide's
+            # handling of file names
+            qis_mod_path = qis_abs_path.replace(os.sep, '.')[:-len('.py')]
+            # users cannot use '/' in filename
+            qis_mod_path = qis_mod_path.replace("/", '.')
+
+            qis_class_name = "reload and rebuild no name"
+            mymodule = importlib.import_module(qis_mod_path)
```
Comment (priti-ashvin-shah-ibm) [inline qiskit_metal/designs/design_base.py:439]: my_module instead of mymodule
To follow naming convention.

File: requirements.txt [hunk 7]
```diff
@@ -12,8 +12,6 @@ pandas
 pint
 pyEPR-quantum>=0.8.4.6
 Pygments
-pyqode.python==2.12.1
```
Comment (priti-ashvin-shah-ibm) [inline requirements.txt:15]: Yeah!

File: qiskit_metal/_gui/widgets/qlibrary_display/delegate_qlibrary.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+        """
+        super().__init__(parent)
+        self.source_model_type = QFileSystemLibraryModel
+        self.is_dev_mode = False
+
+    # https://www.youtube.com/watch?v=v6clAW6FmcU
```
[the remark this answers is not in the document]
Comment (grace-harper-ibm) [reply qiskit_metal/_gui/widgets/qlibrary_display/delegate_qlibrary.py:47]: I saw it as similar to citing StackOverflow but I can remove it 👍

**Comment by priti-ashvin-shah-ibm** [comment]
Spent some time offline with @grace-harper-ibm discussing "developer mode" feature of updating a qcomponent after they have already been imported in QDesign.  

Hot-topic: When a qcomponent  has been updated on hard-drive, but was  already imported into Metal, the user experience was discussed.   I suggest that a warning be given to user, if an user wants the updated import, they should delete all the qcomponents in the design and add them again.  This could be done manually by adding them in the GUI, or through cells which have the qcomponents in the notebook.  Otherwise, pins, connections, and qgeometry tables may become outdated. 

Possible user experience in GUI: If a component has been added to design, the GUI will indicate if the component has been updated on the hard-drive, in developer mode, the tree in qlibrary will have red for files there were edited.  The only way to use the updated qcomponent, is to delete the components in QDesign and then add them back in with updated imports.

**Comment by ThomasGM4** [comment]
> 
> 
> Spent some time offline with @grace-harper-ibm discussing "developer mode" feature of updating a qcomponent after they have already been imported in QDesign.
> 
> Hot-topic: When a qcomponent has been updated on hard-drive, but was already imported into Metal, the user experience was discussed. I suggest that a warning be given to user, if an user wants the updated import, they should delete all the qcomponents in the design and add them again. This could be done manually by adding them in the GUI, or through cells which have the qcomponents in the notebook. Otherwise, pins, connections, and qgeometry tables may become outdated.
> 
> Possible user experience in GUI: If a component has been added to design, the GUI will indicate if the component has been updated on the hard-drive, in developer mode, the tree in qlibrary will have red for files there were edited. The only way to use the updated qcomponent, is to delete the components in QDesign and then add them back in with updated imports.
> 
> @ThomasGM4 thoughts?

I think the delete / over_write would be required, since one couldn't presume how extensive the component has been changed (in theory they could completely modify the options, pins, qgeometry, etc.). So not just a rebuild probably.

**Pull request closed without merging**